### PR TITLE
Combine unit test creation

### PR DESCRIPTION
**Pull request by streeve**
Moved unit test and backend test files to new directory `cmake/test_harness` and combined unit test macros between core and Cajita

### CODE REVIEW

File: core/test_harness/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+# BSD 3-clause license. For the licensing terms see the LICENSE file in    #
+# the top-level directory.                                                 #
+#                                                                          #
+# SPDX-License-Identifier: BSD-3-Clause                                    #
+############################################################################
+
+set(GTEST_SOURCE_DIR ${CMAKE_SOURCE_DIR}/gtest)
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DGTEST_HAS_PTHREAD=0")
+
+include_directories(${GTEST_SOURCE_DIR})
```
Comment (dalg24) [inline core/test_harness/CMakeLists.txt:15]:
```suggestion
```

File: core/test_harness/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+add_library(cabana_gtest ${GTEST_SOURCE_DIR}/gtest/gtest-all.cc)
+set_target_properties(cabana_gtest PROPERTIES
+    CXX_STANDARD 11
+    CXX_STANDARD_REQUIRED YES
+    CXX_EXTENSIONS NO
```
Comment (dalg24) [inline core/test_harness/CMakeLists.txt:20]:
```suggestion
    CXX_EXTENSIONS NO
    COMPILE_DEFINITIONS GTEST_HAS_PTHREAD=0
    INCLUDE_DIRECTORIES ${CMAKE_SOURCE_DIR}/gtest
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -99,6 +99,8 @@ Cabana_add_dependency( PACKAGE Heffte VERSION 2.0.0 )
 # enable unit tests
 option(Cabana_ENABLE_TESTING "Build tests" OFF)
 if(Cabana_ENABLE_TESTING)
+  set(TEST_HARNESS_DIR ${CMAKE_SOURCE_DIR}/core/test_harness)
```
Comment (sslattery) [inline CMakeLists.txt:102]: I thought we wanted the test harness to live in `Cabana/cmake/test_harness`?

File: cajita/unit_test/CMakeLists.txt [hunk 4]
```diff
@@ -110,9 +32,14 @@ set(MPI_TESTS
   SplineEvaluation3d
   SplineEvaluation2d
   Interpolation
+  Parallel
 # BovWriter
   )
 
+if(Kokkos_ENABLE_OPENMPTARGET) #FIXME_OPENMPTARGET
```
Comment (sslattery) [inline cajita/unit_test/CMakeLists.txt:39]: Is the issue here the hierarchical parallelism?

File: core/test_harness/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+# All rights reserved.                                                     #
+#                                                                          #
+# This file is part of the Cabana library. Cabana is distributed under a   #
+# BSD 3-clause license. For the licensing terms see the LICENSE file in    #
+# the top-level directory.                                                 #
+#                                                                          #
+# SPDX-License-Identifier: BSD-3-Clause                                    #
+############################################################################
+
+set(GTEST_SOURCE_DIR ${CMAKE_SOURCE_DIR}/gtest)
```
Comment (sslattery) [inline core/test_harness/CMakeLists.txt:12]: Until #365 is resolved the gtest source should move into `Cabana/cmake/test_harness` as well.

**Review by sslattery — CHANGES_REQUESTED** [review]
I think this looks good. My main comment is I thought we were putting this into the top level `cmake/` folder.

File: cajita/unit_test/CMakeLists.txt (hunk 4 above)
Comment (streeve) [reply cajita/unit_test/CMakeLists.txt:39]: Yes, but we could handle it as discussed #331

File: CMakeLists.txt (hunk 3 above)
Comment (streeve) [reply CMakeLists.txt]: You're right - moved

File: core/test_harness/CMakeLists.txt (hunk 5 above)
Comment (streeve) [reply core/test_harness/CMakeLists.txt]: Moved

**Comment by streeve** [comment]
@sslattery I think this is ready if you're okay with it

File: cajita/unit_test/CMakeLists.txt [hunk 6]
```diff
@@ -9,94 +9,16 @@
...
-        else()
-          add_test(NAME ${_target} COMMAND ${_target} ${gtest_args})
-        endif()
-      endif()
-    endforeach()
-  endforeach()
-endmacro()
+include(${TEST_HARNESS_DIR}/test_harness.cmake)
 
 set(SERIAL_TESTS
```
Comment (sslattery) [inline cajita/unit_test/CMakeLists.txt:14]: This would be for a different PR but I wonder if any of the Cajita tests can be `NO_BACKEND`

**Review by sslattery — APPROVED** [review]
